Resurrect ddclient

bjornfor opened the pull request:
## Description of changes

Upstream has made a new release, so we can bring back ddclient in nixpkgs now.

- Revert "ddclient: remove package and module on upstream maintainer request"
- ddclient: update maintainers: SuperSandro2000 -> bjornfor
- ddclient: 3.10.0 -> 3.11.0_1

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

bjornfor committed:
Revert "ddclient: remove package and module on upstream maintainer request"

This reverts commit d35df28f65208764f6f94ba330c98615d95b934c.

Upstream ddclient has new maintainers and made a new release (update in
next commit(s)).

bjornfor committed:
ddclient: update maintainers: SuperSandro2000 -> bjornfor

SuperSandro2000 doesn't want to maintain this anymore, as per
commit d35df28f65208764f6f94ba330c98615d95b934c
("ddclient: remove package and module on upstream maintainer request").

bjornfor committed:
ddclient: 3.10.0 -> 3.11.0_1

As per the changelog[1], ddclient uses curl instead of perl
IO::Socket::IP and IO::Socket::SSL now.

[1] https://github.com/ddclient/ddclient/blob/d4f9816a6ad86b64caf77af3a15770ca1c1619da/ChangeLog.md#2023-10-15-v3110_1

bjornfor commented:
Result of `nixpkgs-review pr 261241` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ddclient</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/tools/networking/ddclient/default.nix:
```diff
@@ -0,0 +1,45 @@
+  installPhase = ''
+    runHook preInstall
+
+    # patch sheebang ddclient script which only exists after buildPhase
+    preConfigure
```
Maybe worth to investigate if this could be made cleaner

SuperSandro2000 commented on nixos/modules/services/networking/ddclient.nix:
```diff
@@ -0,0 +1,234 @@
+      serviceConfig = {
+        DynamicUser = true;
+        RuntimeDirectoryMode = "0700";
+        inherit RuntimeDirectory;
+        inherit StateDirectory;
```
```suggestion
        inherit RuntimeDirectory StateDirectory;
```
nit

SuperSandro2000 commented on nixos/modules/services/networking/ddclient.nix:
```diff
@@ -0,0 +1,234 @@
+        inherit StateDirectory;
+        Type = "oneshot";
+        ExecStartPre = "!${pkgs.writeShellScript "ddclient-prestart" preStart}";
+        ExecStart = "${lib.getBin cfg.package}/bin/ddclient -file /run/${RuntimeDirectory}/ddclient.conf";
```
```suggestion
        ExecStart = "${lib.getExe cfg.package} -file /run/${RuntimeDirectory}/ddclient.conf";
```
FYI

bjornfor replied:
Updated.

SuperSandro2000 commented on nixos/modules/services/networking/ddclient.nix:
```diff
@@ -0,0 +1,234 @@
+        default = "dyndns2";
+        type = str;
+        description = lib.mdDoc ''
+          Protocol to use with dynamic DNS provider (see https://sourceforge.net/p/ddclient/wiki/protocols).
```
This link is old and not updated anymore

SuperSandro2000 commented on nixos/modules/services/networking/ddclient.nix:
```diff
@@ -0,0 +1,234 @@
+  cfg = config.services.ddclient;
+  boolToStr = bool: if bool then "yes" else "no";
+  dataDir = "/var/lib/ddclient";
+  StateDirectory = builtins.baseNameOf dataDir;
+  RuntimeDirectory = StateDirectory;
```
I am not sure if this variability does not decrease readability a lot.

Majiir reviewed (approved):
Changes LGTM. Verified the revert and built the package.

bjornfor committed:
ddclient: don't run preConfigure in installPhase

Running preConfigure is done to fix the perl shebang in ddclient from
plain perl to perl with modules. But there's a less ugly way: pass the
correct perl to ddclient at configure time.

bjornfor committed:
ddclient: set meta.mainProgram

To silence warning from lib.getExe.

bjornfor committed:
nixos/ddclient: simplify with lib.getExe

bjornfor commented:
Added small improvements (new commits) and rebased to resolve merge conflict.

bjornfor commented:
> Oops, I noticed I had wrongly removed the JSON support (perl module). Drafting until that's fixed.

Fixed now. Sorry for the noise.
